### PR TITLE
version field selector field names in the client

### LINKED ISSUE #6495: kubelet node watch should work in all api versions
My #6423 broke it for v1beta3, then #6442 fixed it for v1beta3 and broke it for v1beta1. I'm working on a fix.

### PR DESCRIPTION
**Pull request by lavalamp**
Fixes #6495.

This is slightly hacky but at least it does the computation in a single place.

### CODE REVIEW

**Comment by lavalamp** [comment]
@quinton-hoole this may be relevant to your interests

**Comment by quinton-hoole** [comment]
Thanks.  I'd rather have someone with better context than me review it.

File: pkg/client/request.go [hunk 1]
```diff
@@ -248,36 +248,85 @@ func (r *Request) RequestURI(uri string) *Request {
...
+var fieldMappings = versionToResourceToFieldMapping{
+	"v1beta1": resourceTypeToFieldMapping{
+		"nodes": clientFieldNameToAPIVersionFieldName{
+			ObjectNameField: "name",
+		},
+		"minions": clientFieldNameToAPIVersionFieldName{
+			ObjectNameField: "name",
+		},
+		"pods": clientFieldNameToAPIVersionFieldName{
+			PodSpecHost: "DesiredState.Host",
```
Comment (bgrant0607) [inline pkg/client/request.go]: Maybe just call it PodHost instead of PodSpecHost? I don't have a strong opinion, though.

**Comment by bgrant0607** [comment]
LGTM. The real solution would be actual versioned clients.

**Comment by vishh** [comment]
LGTM. @lavalamp: Once you address Brian's comment, I will merge this PR

**Comment by lavalamp** [comment]
name changed, PTAL

**Comment by bgrant0607** [comment]
LGTM.

**Comment by smarterclayton** [comment]
This is really hacky.... client/request.go is really not supposed to be tied to any particular object scheme.  Why isn't this on the server side?

**Comment by smarterclayton** [comment]
We have some tools now that we didn't have before around rest handler that should be able to (begin) to solve this problem.  Query parameters from a given api version are deserialized into a versioned object which is converted to the internal form.  GET allows you to provide your own object, but LIST (right now) uses an object which is generic across all resources.  Since field name conversion is tied to a type (field names can't be generically converted) we probably need to give LIST the ability to provide a per Storage query object deserializer.  The one danger there is that it leads to lots of copying (PodListOptions, ServiceListOptions, etc) along with that control.  However, if we eventually want to expose the actual field names that are supported we could do so in the description on that type

The other option would be to add an interface for field transformation per Storage object.  That would probably be the cleanest - the signature could be `func TransformFields(field fields.Selector, version string) fields.Selector` that handled versioning, or it could be a method that returned a map `func FieldTransformations() map[string]map[string]string`.  The latter would be more desirable because we could use that in swagger to list the acceptable values for the fields.

**Comment by smarterclayton** [comment]
Actually, why does rest handler not handle this properly on the server today?  We should be supporting the legacy field name values in the scheme.

**Comment by smarterclayton** [comment]
Oh... it does... on the new Watch.  But we haven't converted the old code to use the new watch yet.  I'll do that now, and then we can remove this.

**Comment by smarterclayton** [comment]
And..... now I see what this is doing.  That was fun.  Never mind....

**Comment by smarterclayton** [comment]
It's causing this to show up in other places.

```
$ kubectl describe svc kubernetes
W0418 20:50:31.002292    3122 request.go:289] field selector: v1beta1 - events - involvedObject.id - kubernetes: need to check if this is versioned correctly.
W0418 20:50:31.002954    3122 request.go:289] field selector: v1beta1 - events - involvedObject.namespace - default: need to check if this is versioned correctly.
W0418 20:50:31.002991    3122 request.go:289] field selector: v1beta1 - events - involvedObject.kind - Service: need to check if this is versioned correctly.
W0418 20:50:31.003005    3122 request.go:289] field selector: v1beta1 - events - involvedObject.uid - 752aeef4-e60c-11e4-8626-080027c5bfa9: need to check if this is versioned correctly.
Name:			kubernetes
Labels:			component=apiserver,provider=kubernetes
Selector:		<none>
IP:			10.0.0.2
Port:			<unnamed>	443/TCP
Endpoints:		10.0.2.15:6443
Session Affinity:	None
No events.
```

**Comment by bgrant0607** [comment]
Should we stop using the query parameters and just change to ListOptions instead?

**Comment by smarterclayton** [comment]
Possibly.  However, we still have to have per kind field filters on both client and server.  We could call convert field labels from the client as well until we switch the client to using versioned objects.

**Pull request merged**